src/main/core/sim_config.rs

```diff
@@ -342,21 +342,35 @@ fn build_process(proc: &ProcessOptions) -> anyhow::Result<Vec<ProcessInfo>> {
             .with_context(|| format!("Failed to parse arguments: {x}"))?,
     };
 
-    // perform shell expansion
-    let plugin = tilde_expansion(proc.path.to_str().unwrap());
-
-    // set argv[0] as the user-provided expanded string, not the canonicalized version
-    args.insert(0, plugin.clone().into());
-
-    // get the full canonical path
-    let plugin = plugin
-        .canonicalize()
-        .with_context(|| format!("Failed to canonicalize plugin path '{}'", plugin.display()))?;
+    let proc_path = proc.path.to_str().unwrap();
+
+    // Check if the path has no '/' where it is likely to be a executable that should be sourced from $PATH
+    // TODO: In next breaking release, if likely_not_path is true, source from $PATH by default and remove deprecation warning below. Also make sure to fix all the broken tests that will result.
+    let likely_not_path = proc_path.chars().find(|c|*c == '/').is_none();
+    
+    let expanded_path = tilde_expansion(proc_path);
+
+    let plugin = match expanded_path
+        .canonicalize().with_context(|| format!("Failed to canonicalize plugin path '{}'", expanded_path.display())) {
+        Ok(plugin) => { // Warn that relative path "thing" should be changed to "./thing" 
+            if likely_not_path { log::warn!("The relative path \"{0}\" will be deprecated in favor of \"./{0}\"", proc_path) }
+            plugin
+        }
+        Err(_) if likely_not_path => {
+            let plugin = which::which(proc_path).with_context(|| format!("Failed to resolve \"{proc_path}\" from $PATH"))?;
+            log::info!("Resolved process path \"{proc_path}\" to \"{}\"", plugin.display());
+            plugin
+        }
+        Err(err) => Err(err)?,
+    };
 
     // verify that the path is a file and is executable
     verify_plugin_path(&plugin)
         .with_context(|| format!("Failed to verify plugin path '{}'", plugin.display()))?;
 
+    // set argv[0] as the user-provided expanded string, not the canonicalized version
+    args.insert(0, expanded_path.into());
```

Comment:
Am I reading correctly that we ultimately always set `argv[0]` the same way we were before - to the result of `tilde_expansion(proc.path)`?

I can see  how that might end up mostly working, since we ultimately call `execvpe`, which does its own PATH-based search. I wouldn't have thought that mattered since we're not setting the `PATH` in the child process's environment, but a) apparently it searches some standard locations anyway, possibly including the current directory, and b) the docs aren't clear whether it looks at the PATH in the `env` being passed to the function, or in the caller's environment.

Since this works in nix, where the binaries aren't in the "usual" locations, I guess it must be using the caller's environment?

I think we'll ultimately want to change how this works a bit to do all of the search and canonicalization ourselves, log the result, and pass the fully canonicalized path to `execvpe`. Otherwise there's a danger that our canonicalization/search is subtly different from what `execvpe` does, and we end up executing a different binary than we've logged. Are you interested in doing that, or would you rather this PR just does the bare minimum to fix the tests under nix?

Reply:
I was a bit confused about what was going on here.

We were *already* passing the canonicalized path separately from argv[0].

The reason the current directory was being "searched" before, is that we were using `std::fs::canonicalize`, which interprets all non-absolute paths as relative to the current directory.

I think what you had here may have been correct, but I think it'd be a bit clearer and more robust to be more explicit about comparing the results  of the "old" and "new" ways. I pushed a commit to do that. Wdyt?

Reply:
The code looks *beautiful* and all the tests seem to pass just fine :rocket:
Looking at the log file though, 

> 00:00:00.109174 [467600:shadow] n/a [WARN] [n/a] [sim_config.rs:374] [shadow_rs::core::sim_config] Expanded path "test-determinism" only via legacy fallback search, which implicitly searched current working dir. Consider prefixing with './'
> 00:00:00.109216 [467600:shadow] n/a [INFO] [n/a] [sim_config.rs:384] [shadow_rs::core::sim_config] Resolved binary path "test-determinism" to "/home/zyansheep/Repos/shadow/build/src/test/determinism/test-determinism"

This part seems like it works just fine

> 00:00:00.008566 [470057:shadow] n/a [INFO] [n/a] [sim_config.rs:384] [shadow_rs::core::sim_config] Resolved binary path "true" to "/nix/store/1xqr1166ji7xy80dp8d2nn1a6x8m6m5w-coreutils-9.1/bin/coreutils"

But shouldn't this part have a `/bin/coreutils/true` at the end?

Edit: nvm just figure it out, coreutils represents all the coreutils programs in one binary...

Reply:
Looks good to me!